File: frontend/src/app/core/services/email.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpParams } from '@angular/common/http';
import { ApiService } from './api.service';
import { Email, EmailPage, EmailSyncResult, ComposeEmail, ComposeEmailResponse, DraftAttachment } from '../../models/email.model';

/**
 * Provides account-scoped email operations including paginated listing with filters,
 * read/star toggling, IMAP synchronization, reprocessing, and attachment downloads.
 */
@Injectable({ providedIn: 'root' })
export class EmailService {
  private readonly api = inject(ApiService);

  private basePath(accountId: string) {
    return `/email-accounts/${accountId}/emails`;
  }

  list(accountId: string, options?: {
    page?: number; size?: number; folder?: string; query?: string; isRead?: boolean;
    dateFrom?: string; dateTo?: string;
    categoryId?: string; processed?: boolean; automationId?: string;
  }) {
    let params = new HttpParams();
    if (options?.page != null) params = params.set('page', options.page);
    if (options?.size != null) params = params.set('size', options.size);
    if (options?.folder) params = params.set('folder', options.folder);
    if (options?.query) params = params.set('query', options.query);
    if (options?.isRead != null) params = params.set('isRead', options.isRead);
    if (options?.dateFrom) params = params.set('dateFrom', options.dateFrom);
    if (options?.dateTo) params = params.set('dateTo', options.dateTo);
    if (options?.categoryId) params = params.set('categoryId', options.categoryId);
    if (options?.processed != null) params = params.set('processed', options.processed);
    if (options?.automationId) params = params.set('automationId', options.automationId);
    return this.api.get<EmailPage>(this.basePath(accountId), { params });
  }

  get(accountId: string, emailId: string) {
    return this.api.get<Email>(`${this.basePath(accountId)}/${emailId}`);
  }

  markRead(accountId: string, emailId: string, read: boolean) {
    return this.api.patch<Email>(`${this.basePath(accountId)}/${emailId}/read`, { read });
  }

  toggleStar(accountId: string, emailId: string) {
    return this.api.patch<Email>(`${this.basePath(accountId)}/${emailId}/star`, {});
  }

  reprocess(accountId: string, emailId: string) {
    return this.api.post<Email>(`${this.basePath(accountId)}/${emailId}/reprocess`, {});
  }

  sync(accountId: string) {
    return this.api.post<EmailSyncResult>(`${this.basePath(accountId)}/sync`, {});
  }

  downloadAttachment(accountId: string, emailId: string, index: number) {
    return this.api.getBlob(`${this.basePath(accountId)}/${emailId}/attachments/${index}`);
  }

  /** Moves an email to Trash; deleting one already in Trash removes it permanently. */
  deleteEmail(accountId: string, emailId: string) {
    return this.api.delete<void>(`${this.basePath(accountId)}/${emailId}`);
  }

  /** Restores a trashed email back to its original folder. */
  restoreEmail(accountId: string, emailId: string) {
    return this.api.post<void>(`${this.basePath(accountId)}/${emailId}/restore`, {});
  }

  /** Permanently deletes all trashed emails in the mailbox. */
  emptyTrash(accountId: string) {
    return this.api.delete<{ deleted: number }>(`${this.basePath(accountId)}/trash`);
  }

  send(accountId: string, request: ComposeEmail) {
    return this.api.post<ComposeEmailResponse>(`${this.basePath(accountId)}/send`, request);
  }

  saveDraft(accountId: string, request: ComposeEmail) {
    return this.api.post<ComposeEmailResponse>(`${this.basePath(accountId)}/drafts`, request);
  }

  updateDraft(accountId: string, draftId: string, request: ComposeEmail) {
    return this.api.put<ComposeEmailResponse>(`${this.basePath(accountId)}/drafts/${draftId}`, request);
  }

  deleteDraft(accountId: string, draftId: string) {
    return this.api.delete<void>(`${this.basePath(accountId)}/drafts/${draftId}`);
  }

  uploadAttachment(accountId: string, draftId: string, file: File) {
    const formData = new FormData();
    formData.append('file', file);
    return this.api.post<DraftAttachment>(`${this.basePath(accountId)}/drafts/${draftId}/attachments`, formData);
  }

  deleteAttachment(accountId: string, draftId: string, attachmentId: string) {
    return this.api.delete<void>(`${this.basePath(accountId)}/drafts/${draftId}/attachments/${attachmentId}`);
  }

  listAttachments(accountId: string, draftId: string) {
    return this.api.get<DraftAttachment[]>(`${this.basePath(accountId)}/drafts/${draftId}/attachments`);
  }
}

File: backend/src/main/java/com/postwerk/service/EmailService.java
package com.postwerk.service;

import com.postwerk.dto.EmailCategoryItem;
import com.postwerk.dto.EmailListResponse;
import com.postwerk.dto.EmailResponse;
import com.postwerk.dto.EmailSyncResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

/**
 * Service interface for email operations including paginated listing with advanced filtering,
 * read/star toggling, category assignment, IMAP sync triggering, and attachment downloads.
 *
 * @since 1.0
 */
public interface EmailService {

    Page<EmailListResponse> list(UUID organizationId, UUID accountId, String folder, String query, Boolean isRead,
                                  Instant dateFrom, Instant dateTo,
                                  UUID categoryId, Boolean processed, UUID automationId,
                                  Pageable pageable);

    EmailResponse getById(UUID organizationId, UUID actingUserId, UUID accountId, UUID emailId, String ipAddress);

    EmailResponse markRead(UUID organizationId, UUID accountId, UUID emailId, boolean read);

    EmailResponse toggleStar(UUID organizationId, UUID accountId, UUID emailId);

    EmailResponse assignCategories(UUID organizationId, UUID accountId, UUID emailId, List<EmailCategoryItem> categories);

    EmailResponse reprocess(UUID organizationId, UUID accountId, UUID emailId);

    EmailSyncResponse sync(UUID organizationId, UUID actingUserId, UUID accountId, String ipAddress);

    Object[] downloadAttachment(UUID organizationId, UUID accountId, UUID emailId, int attachmentIndex);

    /**
     * Moves an email to Trash (sets {@code trashedAt}). If the email is already in Trash, it is
     * permanently (soft-)deleted instead.
     */
    void deleteEmail(UUID organizationId, UUID accountId, UUID emailId);

    /** Restores a trashed email back to its original folder (clears {@code trashedAt}). */
    void restoreEmail(UUID organizationId, UUID accountId, UUID emailId);

    /** Permanently (soft-)deletes all trashed emails in the mailbox. Returns the number removed. */
    int emptyTrash(UUID organizationId, UUID accountId);
}
